### PR TITLE
Keep shared skills within Tau load limits

### PR DESCRIPTION
**Pull request by douglaz**
## Summary

- split three oversized workflows into exact-name internal companion skills so Tau can load every procedure within its 65,536-byte source limit
- teach selective installation and target-local upgrade repair about companion dependencies
- pin Tau source/description limits, companion contracts, installer self-reexec, and upgrade edge cases in install.test
- harden moved backlog-drain and PR-merge procedures while preserving their owning workflow entry points

## Tradeoffs

- internal companions are model-loadable by exact name and set user-invocable: false
- explicit selective installs receive dependency closure; upgrade reconciliation repairs one direct hop per target to avoid silently adding standalone workflows
- a pre-self-reexec installer may require the documented one-time repeated selective update

## Verification

- ./install.test — passed 25, failed 0
- ./skills/pr-with-codex-bot-review/scripts/bot-gate.test — passed 124, failed 0
- ./skills/drive/scripts/drive-status.test — passed 70, failed 0
- git diff --cached --check — clean before commit

## Local review

Codex gpt-5.6-sol at xhigh and Claude Opus 5 at high effort reviewed iteratively through 21 confirmation rounds. Accepted findings were fixed and covered where practical.

Per explicit disposition, the final three findings are filed rather than included in this PR:

- #61 — zero-argument installer behavior on Bash 4.0–4.3
- #62 — modified secondary checkout can reexec shared-clone code
- #63 — quoted YAML-equivalent companion names are rejected

## GitHub review

- CI: pending
- chatgpt-codex-connector: pending
- CodeRabbit: pending if configured

## Summary by CodeRabbit

* **New Features**
  * Added automated companion-skill discovery, installation, upgrade, and repair.
  * Added workflows for delegated review edits, backlog processing, and safe pull request merging.
  * Added validation for skill descriptions, required files, links, and recovery guidance.

* **Documentation**
  * Clarified gate outcomes, merge-blocking statuses, installation behavior, and companion-skill usage.
  * Updated workflow references and evaluation criteria.

* **Tests**
  * Expanded coverage for installer updates, selective installation, companion recovery, validation, and safety checks.

### CODE REVIEW

**Comment by douglaz** [comment]
@codex review

**Comment by douglaz** [comment]
@codex review

File: install.sh [hunk 1]
```diff
@@ -83,14 +93,165 @@ esac
...
-  local base="$1"
+  local base="$1" d
   local found=()
   for d in "$base"/skills/*/; do
+    # Matches .gitignore: eval-creator `*-workspace` scaffolds are not skills.
+    [[ ${d%/} == *-workspace ]] && continue
     [[ -f "$d/SKILL.md" ]] && found+=("$(basename "$d")")
   done
   printf '%s\n' "${found[@]}"
 }
```
[the remark this answers is not in the document]
Comment (douglaz) [reply install.sh:104]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: install.test [hunk 2]
```diff
@@ -160,5 +180,332 @@ if [[ $rc -eq 0 ]]; then fail "$t" "exit 0 for a skill that does not exist"
...
+printf '%s\n' '#!/usr/bin/env bash' 'cp "$SELF_UPDATE_SOURCE" "$SELF_UPDATE_TARGET.next" && mv "$SELF_UPDATE_TARGET.next" "$SELF_UPDATE_TARGET"' > "$b/bin/git"
+chmod +x "$b/bin/git"
+SELF_UPDATE_SOURCE="$b/clone/install.sh.updated" SELF_UPDATE_TARGET="$b/clone/install.sh" \
+  HOME="$b/home" CODEX_HOME="$b/codexhome" PATH="$b/bin:$PATH" \
+  bash "$b/clone/install.sh" --dir "$b/clone" --target claude alpha >/dev/null; rc=$?
+observed=$(< "$b/home/reexec-observed")
+if [[ $rc -ne 0 ]]; then fail "$t" "exit $rc"
+elif [[ "$observed" != "--dir $b/clone --target claude alpha" ]]; then
+  fail "$t" "updated installer observed wrong argv: $observed"
+else pass "$t"; fi
```
[the remark this answers is not in the document]
Comment (douglaz) [reply install.test:215]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/drive/SKILL.md [hunk 3]
```diff
@@ -1,22 +1,16 @@
...
+  Drive a software goal across lifecycle phases until its finish condition is met.
+  Use when work spans spec, beads, build, tests, reviewer convergence, PR/merge, or
+  multiple backlog items—even if the user never says "drive". Trigger on "drive
+  this", "end to end", "from idea/spec to merged", "run the whole pipeline",
+  "until it's done", "drain the backlog", "don't stop and ask me between steps",
+  "I'll check back later", feature/epic/milestone goals, project status, or resuming
+  half-finished work. It routes to specialist skills and enforces
+  evidence at every transition. Skip for a single bounded action, one-off edit,
+  question, or debugging/ops task. Prefer Drive over invoking phase skills directly
+  whenever the work crosses phases.
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/drive/SKILL.md:13]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/multi-reviewer-loop-delegating-edits/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  - **Use the harness edit tool, not `sed -i` or `str.replace`**, which report success
+    on zero matches. If an edit must be scripted, assert the target exists before
+    replacing, then grep the file for both the new text and the *absence* of the old.
+    This is the repo's standing edit contract (`agents-md`), and it does not travel to
+    a fresh context on its own.
+  - **Add no *unrequested* mechanism** — not "nothing new". A verified finding may
+    genuinely require a lock, a validation, or a transaction; what is forbidden is
+    anything beyond the finding. Scope conflicts come back to you to adjudicate, not
+    resolved by the implementer.
+  - Plus your environment guards.
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/multi-reviewer-loop-delegating-edits/SKILL.md:73]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/multi-reviewer-loop-delegating-edits/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    needs no filtering: `git apply --include` takes a **pattern**, and both a
+    bare directory and a literal `a[1].txt` silently match no hunks while exiting 0 — after
+    step 2 has already reverted those paths, which turns a rollback into a loss that reports
+    success. One combined `git diff HEAD` flattens the layers, so an `MM` path
+    comes back ` M` and a later commit carries hunks the user left out. Both flags matter:
+    without `--binary` a modified binary records only "Binary files differ" and will not
+    apply; `--no-textconv` because the conversion is not what is on disk.
+  - `git status --porcelain -z`, NUL-delimited. Porcelain quotes paths that need it, so a
+    `cut`-based parse hands back a literal `"caf\303\251.py"` that does not exist. This
+    is also what restores **intent-to-add** entries (` A `) afterwards:
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/multi-reviewer-loop-delegating-edits/SKILL.md:108]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/multi-reviewer-loop-delegating-edits/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  source staged as deleted and the replay then fails); contents inside a dirty submodule;
+  anything with a byte-transforming `.gitattributes` entry (`git check-attr --all` —
+  `filter`, `text`, `eol`, `working-tree-encoding`); text files under `core.autocrlf`, in **either**
+  `true` or `input` — both convert, with no attribute for `check-attr` to report.
+  Measured: under `true` a mixed CRLF/LF file replayed as all-CRLF from a patch that
+  applied cleanly; under `input` a CRLF worktree against an LF index gave an empty patch
+  for a file `git status` called modified, and the rollback then rewrote it to LF;
+  and files marked `assume-unchanged`,
+  where all three captures *succeed while recording nothing* and the revert then replaces
+  the bytes with `HEAD`. That list is the cases known to hit the rule, not the boundary.
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/multi-reviewer-loop-delegating-edits/SKILL.md:134]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/multi-reviewer-loop-delegating-edits/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+     already reverted. Path-restricted capture is also what keeps the replay from touching
+     unrelated hunks that step 2 deliberately left applied — replaying those would abort
+     the whole patch with `patch does not apply` before any delegated path is restored.
+     **Skip an empty patch file** (or pass `--allow-empty`): an
+     unstaged-only edit leaves the staged layer empty and `git apply` exits 128 on it —
+     after the revert, so the layer holding the user's work is never applied
+  4. restore each untracked backup, clearing its destination first
+  5. re-run `git add -N` for every path the status snapshot recorded as `" A "` — a leading
+     space, then `A`; `"A "` is an ordinary staged addition and must not be re-added
+
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/multi-reviewer-loop-delegating-edits/SKILL.md:158]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/orchestrating-with-rb-lite/references/harden-until-clean.md [hunk 8]
```diff
@@ -369,15 +370,16 @@ field-level changes: a full re-serialization with every id on both sides is
...
 If the iteration ran degraded, say which reviewer was missing in that commit
 message. Months later it explains why iteration N looks thin.
 
 ## 4. Drain the beads
 
-Hand off to the **Backlog-drain workflow** in the main skill and follow it
-exactly: one bead, one branch, one rb-lite run, local gates, one **work** PR, one
+Hand off to exact companion skill
+[`rb-lite-backlog-drain`](../../rb-lite-backlog-drain/SKILL.md) and follow it exactly:
+one bead, one branch, one rb-lite run, local gates, one **work** PR, one
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/orchestrating-with-rb-lite/references/harden-until-clean.md:382]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/pr-with-codex-bot-review-merge/SKILL.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+if [ "$GATE_RC" -eq 0 ]; then
+  [ "$(printf %s "$GATE_JSON" | jq -r .verdict)" = "NO_PENDING_EVIDENCE" ] \
+    || { echo "gate verdict is not NO_PENDING_EVIDENCE"; exit 1; }
+  REVIEWED_TIP=$(printf %s "$GATE_JSON" | jq -er .tip) \
+    || { echo "gate JSON has no tip"; exit 1; }
+elif [ "$GATE_RC" -eq 4 ]; then
+  GATE_EVIDENCE=$(git rev-parse --git-path "merge-gate-<N>.json") \
+    || { echo "cannot resolve the gate-evidence path"; exit 1; }
+  (umask 077; printf '%s\n' "$GATE_JSON" > "$GATE_EVIDENCE") \
+    || { echo "cannot preserve exit-4 gate JSON"; exit 1; }
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/pr-with-codex-bot-review-merge/SKILL.md:114]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/pr-with-codex-bot-review-merge/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+  [ "$_ST" = "CLOSED" ] && { echo "PR was closed without merging"; exit 1; }
+  sleep 10
+done
+[ "$_ST" = "MERGED" ] || { echo "PR still not merged (state=$_ST) — do NOT proceed"; exit 1; }
+
+# Keep $GATE_JSON: the merge report has to QUOTE it (see
+# the already-loaded owning skill's § 8a; see
+# ../pr-with-codex-bot-review/SKILL.md#8a-the-merge-report-must-quote-the-gate). Do not re-run the gate to
+# produce the quote — the PR is merged by now, so a fresh run describes a different world.
+printf '%s\n' "$GATE_JSON" > "${TMPDIR:-/tmp}/merge-evidence-<N>.json"
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/pr-with-codex-bot-review-merge/SKILL.md:176]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/rb-lite-backlog-drain/SKILL.md [hunk 11]
```diff
@@ -0,0 +1,538 @@
...
+    # this read and the merge widens the window. It cannot reach zero (GitHub has no base
+    # pin), but it is one API call wide.
+    BASE_NOW=$(gh pr view <pr> -R "$R" --json baseRefName -q .baseRefName) \
+      || { echo "cannot re-read the merge target — do NOT merge"; exit 1; }
+    [ -n "$BASE_NOW" ] && [ "$BASE_NOW" = "$BASE" ] \
+      || { echo "the PR was retargeted ($BASE -> ${BASE_NOW:-unknown}) — the panel reviewed against $BASE; re-run the panel"; exit 1; }
+    gh pr merge <pr> -R "$R" --squash --delete-branch --match-head-commit "$MERGING_SHA" \
+      || { echo "merge did not land — do NOT close the bead"; exit 1; }
+
+    # `gh pr merge` returning 0 means the PR was accepted for merging — which, in a repo with a
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/rb-lite-backlog-drain/SKILL.md:273]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/rb-lite-backlog-drain/SKILL.md [hunk 12]
```diff
@@ -0,0 +1,538 @@
...
+    br sync --flush-only || { echo "not persisted"; exit 1; }
+    ```
+
+    **Check for divergence BEFORE the first `br` write.** `br update` auto-flushes, so a
+    hand-edit sitting unstaged in the JSONL is destroyed by the very first mutation — and
+    since neither the index nor `HEAD` holds it, every later diff shows only your intended
+    change, which makes the loss *undetectable*, not merely unrecoverable. Resolve the real
+    path (`br where --json | jq -er .jsonl_path`; `.beads/issues.jsonl` is only the default
+    layout, and a hardcoded path diffs nothing on a `.beads.jsonl` repo — a false
+    all-clear) and run `git status --porcelain` on it before any `br` command.
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/rb-lite-backlog-drain/SKILL.md:376]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.

File: skills/rb-lite-backlog-drain/SKILL.md [hunk 13]
```diff
@@ -0,0 +1,538 @@
...
+    # whenever the marker query above is empty — see below. MERGED-only hid the third
+    # resume state: a work PR opened but not yet landed carries the bead id and no
+    # `bead-closure:` marker, so neither query saw it and the drain re-entered step 6 to
+    # rebuild work that was already open for review. Run separately this would re-fetch,
+    # and a fetch that failed here while the marker query succeeded would silently read as
+    # "nothing merged", the exact collapse this query exists to close.
+    printf '%s\n' "${CLOSURE_PRS[@]}" \
+      | jq -s --arg id "$BEAD_ID" 'add | unique_by(.url) | .[] | select(.state=="MERGED" or .state=="OPEN")
+          | select($id != "" and ((.body // "") | ascii_downcase
+            | test("(^|[^a-z0-9._-])" + ($id|ascii_downcase|gsub("\\.";"\\.")) + "([^a-z0-9._-]|$)")))'
```
[the remark this answers is not in the document]
Comment (douglaz) [reply skills/rb-lite-backlog-drain/SKILL.md:490]: Deferred per maintainer direction and tracked in #65. Resolving this thread as filed, not fixed.
